Enable reconfiguration of the DeviceNode GW

kursatkobya opened the pull request:
Add isConfigured variable to devicenodeparser.h
Add component tests for reconfiguring DeviceNodeGateway

Signed-off-by: kursat <kursat.kobya@pelagicore.com>

kursatkobya committed:
Enable reconfiguration of the DeviceNode GW
Add isConfigured variable to devicenodeparser.h
Add component tests for reconfiguring DeviceNodeGateway

Signed-off-by: kursat <kursat.kobya@pelagicore.com>

tobsan commented on libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.h:
```diff
@@ -70,6 +70,13 @@ class DeviceNodeGateway :
+     *
+     * return true or false regarding configuration status
+     */
+    virtual bool isDeviceConfigured(const std::string deviceName);
```
Shouldn't this be private instead of public?

kursatkobya replied:
This function is needed by gtest to examine the idea of "only update device when it is needed".

frznlogic replied:
Could it be made protected instead in that case?

tobsan commented on libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.cpp:
```diff
@@ -51,38 +51,41 @@ bool DeviceNodeGateway::activateGateway()
             }
+
+            // Mount device in container
+            getContainer()->mountDevice(dev.name);
```
Shouldn't there be some error checking here?

kursatkobya replied:
this part is not new code, but sure.

breakreturn commented on libsoftwarecontainer/component-test/devicenodegateway_componenttest.cpp:
```diff
@@ -97,3 +97,60 @@ TEST_F(DeviceNodeGatewayTest, TestOverwriteDeviceFails) {
+            true}
+));
+
+TEST_P(DynamicConfigurationTests, DynamicDeviceNodeConfiguration) {
```
Needs a comment what the test is about etc.

breakreturn commented on libsoftwarecontainer/component-test/devicenodegateway_componenttest.cpp:
```diff
@@ -97,3 +97,60 @@ TEST_F(DeviceNodeGatewayTest, TestOverwriteDeviceFails) {
+            "[{\"name\": \"/dev/tty0\", \"mode\":  622}]",
+            "[{\"name\": \"/dev/tty0\", \"mode\":  755}]",
+            "/dev/tty0",
+            false},
```
For readability, could you add a space before the opening braces and break the line for the closing braces and fix the indentation so the config strings are indented one level and the braces matches on the outer level?

Like:

testSetup {
    "string",
    "string"
},
testSetup {

etc.

tobsan commented on libsoftwarecontainer/component-test/devicenodegateway_componenttest.cpp:
```diff
@@ -97,3 +97,60 @@ TEST_F(DeviceNodeGatewayTest, TestOverwriteDeviceFails) {
+struct testSetup
+{
+    const std::string firstConfiguration;
+    const std::string secondConfiguration;
```
It would be nice to be able to use more than two configuration stages in the tests, since the implementation is not limited to two afaict.

breakreturn replied:
This is a good point, the quick-launch feature might not require more than two sequences at the moment, but our support for this in general is not so far limited to two and I don't think this should be either.

tobsan reviewed (changes requested):
Looks good, just had some comments. I like the general idea to mark a device as up-to-date or not.

tobsan commented on libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.cpp:
```diff
@@ -51,38 +51,45 @@ bool DeviceNodeGateway::activateGateway()
-        std::string devicePathInContainerOnHost = buildPath(getContainer()->rootFS(), dev.name);
-        std::string deviceParent = parentPath(devicePathInContainerOnHost);
+            std::string devicePathInContainerOnHost = buildPath(getContainer()->rootFS(), dev.name);
+            std::string deviceParent = parentPath(devicePathInContainerOnHost);
```
I just realized that this whole check for `deviceParent` etc only works because we are lucky. The /dev/ directory inside the container is mounted as a tmpfs, and as such, we can't see it from the host side.

@kursatkobya can you update the test to try to mount `/dev/input/event0` into the container? I suspect this should fail (I should've taken care of this when I handled the general bind-mounting issue) because `/dev/input` does not exist inside the container, and the `createDirectory` calls will only create it on the host (not inside the tmpfs where `/dev` in the container resides).

If that is the case, we should update to try to create any needed parent directories inside the container (with a functionjob, I guess).

kursatkobya committed:
Remove deprecated tests from devicenodelogic_unittest

Signed-off-by: kursat <kursat.kobya@pelagicore.com>

kursatkobya committed:
Remove deprecated directory creation and check flow on DeviceNodeGateway

Signed-off-by: kursat <kursat.kobya@pelagicore.com>

frznlogic commented on libsoftwarecontainer/src/gateway/devicenode/devicenodegateway.cpp:
```diff
@@ -51,38 +51,29 @@ bool DeviceNodeGateway::activateGateway()
-        std::string deviceParent = parentPath(devicePathInContainerOnHost);
-
-        // Already existing files can't be converted into directories
-        if (existsInFileSystem(deviceParent) && !isDirectory(deviceParent)) {
```
Are these checks performed somewhere else or why are we removing them?

kursatkobya replied:
Those lines are removed, because we are no longer support creating and mounting new devices.

tobsan replied:
or rather, when we add an existing device, LXC handles creation of any parent directories.

kursatkobya committed:
Add Device class
Add unit tests for Device class to test activation/configuration mentality.

The aim of creating this class was making DeviceNodeGateway more object
oriented and more approppriate for testing units. This commit is changed
the way devices stores in the gateway properties. With this change each
device will parse configuration for themselves and will activate
themselves.

Signed-off-by: kursat <kursat.kobya@pelagicore.com>

tobsan commented on libsoftwarecontainer/src/gateway/devicenode/devicenode.cpp:
```diff
@@ -0,0 +1,94 @@
+{
+}
+
+bool Device::parse(const json_t *element)
```
I would prefer if this function was a static factory function, so that it creates a Device object on successful parse.

kursatkobya replied:
we can overload constructor but i think this way is easier to read.

tobsan commented on libsoftwarecontainer/src/gateway/devicenode/devicenode.h:
```diff
@@ -0,0 +1,69 @@
+    LOG_DECLARE_CLASS_CONTEXT("DN", "Device Node");
+
+public:
+    Device(std::string _name = "", int _mode = -1, bool _condigured = false);
```
I don't think we need defaults for these, do we?

kursatkobya replied:
except iscondigured it would be better to have defaults imo.

tobsan commented on libsoftwarecontainer/src/gateway/devicenode/devicenode.h:
```diff
@@ -0,0 +1,69 @@
+     */
+    bool activate(std::shared_ptr<ContainerAbstractInterface> container);
+
+    std::string name;
```
This should be private with getter/setter if needed.

tobsan commented on libsoftwarecontainer/src/gateway/devicenode/devicenode.h:
```diff
@@ -0,0 +1,69 @@
+     * @brief Holds device mode which consist 3 integer number between 0 and 7 which represents
+     * permissions for owner, groups and all users
+     * */
+    int mode;
```
This should be private with getter/setter if needed.

tobsan commented on libsoftwarecontainer/src/gateway/devicenode/devicenode.h:
```diff
@@ -0,0 +1,69 @@
+    /*
+    * @brief Holds the flag that whether the device is configured or not
+    */
+    bool isConfigured;
```
This should be private with getter/setter if needed.

tobsan commented on libsoftwarecontainer/src/gateway/devicenode/devicenodelogic.cpp:
```diff
@@ -18,33 +18,40 @@
-            [&] (DeviceNodeParser::Device const &d) { return d.name == name; });
+            [&] (std::shared_ptr<Device> const &d) { return d->name == name; });
 
     return item;
```
I think it's a bit weird to return an iterator to a shared_ptr. Can't you return the shared_ptr directly instead?

tobsan commented on libsoftwarecontainer/src/gateway/devicenode/devicenodelogic.cpp:
```diff
@@ -18,33 +18,40 @@
+        if (mode != (*item)->mode) {
+            (*item)->isConfigured = false;
+            (*item)->mode = mode;
+        }
```
If you return the shared_ptr directly, from findDeviceByName, this code would look way better.

tobsan commented on libsoftwarecontainer/unit-test/devicenode_unittest.cpp:
```diff
@@ -0,0 +1,211 @@
+struct testSetup
+{
+    Device firstConfiguration;
+    Device secondConfiguration;
```
As I said earlier, it would be nice to be able to test more than two configuration runs. Perhaps make it a vector instead?

kursatkobya replied:
We can create another test to examine this. But what would have been the perks of having it ? Since this feature is provided by inherited Gateway class, it is actually on/off option. Either a gateway can be activated once or no limitation.

tobsan replied:
I just mean, instead of testing one or two activations, we could test one or more activations.

tobsan reviewed (changes requested):
I really like the re-design of this gateway. However, it looks super weird to hold shared_ptr of Device in DeviceNodeLogic, when the only pointer operations are done inside the logic class itself. That feels completely backwards.

I would propose that the Device class itself can do the whitelisting operations now done in DeviceNodeLogic, or that the lookup of devices in the logic class is changed if possible.

tobsan commented on libsoftwarecontainer/src/gateway/devicenode/devicenode.cpp:
```diff
@@ -0,0 +1,138 @@
+Device::Device(std::string _name, int _mode) :
+        name(_name),
+        mode(_mode),
+        isConfigured(false)
```
Perhaps nitpicking, but you should use m_name, m_mode and m_isConfigured for the members. Then you won't need _name etc.

tobsan commented on libsoftwarecontainer/src/gateway/devicenode/devicenode.cpp:
```diff
@@ -0,0 +1,138 @@
+bool Device::getIsconfigured ()
+{
+    return isConfigured;
+}
```
I would prefer if the getters had the name of the variable, so mode(), name() etc.

tobsan commented on libsoftwarecontainer/unit-test/devicenode_unittest.cpp:
```diff
@@ -0,0 +1,249 @@
+    dev.setMode(testparams.itemMode);
+    dev.calculateDeviceMode(testparams.newMode);
+    ASSERT_EQ(testparams.expectedMode, dev.getMode());
+}
```
These tests perform the same testing as the component test that's added in this PR does, right? Then I would vote for removing the component test. We should add a service test, but that should not be part of this PR imo.

kursatkobya replied:
The component tests additionally performs mount operations. So i think we should remove them after adding some related service tests.

tobsan reviewed (changes requested):
Some new comments.
